File: 11_动态规划/dp分类/路径dp/63. 不同路径 II-障碍物.py
from functools import lru_cache
from typing import List


DIR2 = ((1, 0), (0, 1))


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        @lru_cache(None)
        def dfs(row: int, col: int) -> int:
            if obstacleGrid[row][col] == 1:
                return 0
            if (row, col) == (ROW - 1, COL - 1):
                return 1

            res = 0
            for dr, dc in DIR2:
                nr, nc = row + dr, col + dc
                if 0 <= nr < ROW and 0 <= nc < COL and obstacleGrid[nr][nc] == 0:
                    res += dfs(nr, nc)
            return res

        ROW, COL = len(obstacleGrid), len(obstacleGrid[0])
        res = dfs(0, 0)
        dfs.cache_clear()
        return res

File: 11_动态规划/dp分类/路径dp/63. 不同路径 II-障碍物.ts
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
const uniquePathsWithObstacles = (obstacleGrid: number[][]): number => {
  if (obstacleGrid[0][0] === 1) return 0
  const m = obstacleGrid.length
  const n = obstacleGrid[0].length
  const dp: number[][] = Array.from({ length: m }, () => Array(n).fill(0))

  dp[0][0] = obstacleGrid[0][0] === 1 ? 0 : 1

  for (let i = 1; i < m; i++) {
    if (obstacleGrid[i][0] === 1) continue
    dp[i][0] = dp[i - 1][0]
  }

  for (let j = 1; j < n; j++) {
    if (obstacleGrid[0][j] === 1) continue
    dp[0][j] = dp[0][j - 1]
  }

  for (let i = 1; i < m; i++) {
    for (let j = 1; j < n; j++) {
      if (obstacleGrid[i][j] === 1) continue
      dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
    }
  }

  console.table(dp)
  return dp[m - 1][n - 1]
}

console.log(
  uniquePathsWithObstacles([
    [0, 1],
    [0, 0]
  ])
)
console.log(
  uniquePathsWithObstacles([
    [0, 0],
    [0, 1]
  ])
)
console.log(
  uniquePathsWithObstacles([
    [0, 0],
    [1, 1],
    [0, 0]
  ])
)
// console.log(uniquePathsWithObstacles([[1, 0]]))
// 输出：2
export {}
